### PR TITLE
Tutorial on how to create a basic brick

### LINKED ISSUE #611: "Creating a Brick" tutorial
Walking through my second or so time attempting to create a custom brick, I think it would be a good idea to have a "custom brick" tutorial, in the same spirit as the Theano tutorials that walk you through the creation of a custom op (albeit more high-level).

### PR DESCRIPTION
**Pull request by adbrebs**
Fixes #611 @dmitriy-serdyuk 
This is still work in progress but you can already comment on the structure.

### CODE REVIEW

**Comment by dmitriy-serdyuk** [comment]
The structure looks good.

We test example code, so maybe you'll need to skip some tests, see [doctest](https://docs.python.org/2/library/doctest.html) module.

File: docs/create_your_own_brick.rst [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+methods (strictly speaking, all these methods are optional):
+
+* :meth:`.Brick.__init__`: you should pass by argument the attributes of your
+  bricks. It is also in this method that you should create the potential
+  "children bricks" that belongs to your brick. The initialiazation of the
+  attributes can be lazy as described in a further paragraph.
+* :meth:`you_decide_which_name`: you need to implement a method that actually
+  implements the operation of the brick, taking as arguments the inputs
+  of the brick and returning its outputs. It can have any name and for simple
+  bricks is often named ``apply``. You can decorate it with the
```
Comment (dmitriy-serdyuk) [inline docs/create_your_own_brick.rst]: You **should**, otherwise all blocks magic won't work.

File: docs/create_your_own_brick.rst [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+  :func:`.application` decorator, as explained in the next section.
+* :meth:`.Brick._allocate`: you should implement this method if your brick
+  needs to allocate its parameters.
+* :meth:`.Brick._initialize`: you should implement this method if you need to
+  initialize parameters of your brick.
+* :meth:`.Brick._push_allocation_config`: you should consider overwriting
+  this method if you want to allocate the children bricks in a specific way.
+* :meth:`.Brick._push_initialization_config`: you should consider method if
+  you want to initialize the children bricks in a specific way.
```
Comment (dmitriy-serdyuk) [inline docs/create_your_own_brick.rst]: It worth mentioning here that usually one should just inherit from `Initializable` and it is going to initialize children recursively.

File: docs/create_your_own_brick.rst [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    ...         self.output_dim1 = output_dim1
+    ...         self.output_dim2 = output_dim2
+    ...
+    ...     @property
+    ...     def W1(self):
+    ...         return self.parameters[0]
+    ...
+    ...     @property
+    ...     def W2(self):
+    ...         return self.parameters[0]
```
Comment (dmitriy-serdyuk) [inline docs/create_your_own_brick.rst]: 0 -> 1

File: docs/create_your_own_brick.rst [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    ...         add_role(W, WEIGHT)
+    ...         self.parameters.append(W)
+    ...         self.add_auxiliary_variable(W.norm(2), name='W'+number+'_norm')
+    ...
+    ...     def _allocate(self):
+    ...         self.__allocate(self.input_dim1, self.output_dim1, '1')
+    ...         self.__allocate(self.input_dim2, self.output_dim2, '2')
+    ...
+    ...     def _initialize(self):
+    ...         W1, W2 = self.parameters
```
Comment (dmitriy-serdyuk) [inline docs/create_your_own_brick.rst]: You could use `self.W*` here.

**Comment by dmitriy-serdyuk** [comment]
Can you also add an example on how to apply the brick to theano variables, compile a function and check the result?

**Comment by dmitriy-serdyuk** [comment]
And can you add a See Also link to recurrent stuff? I think we should add a similar tutorial for recurrent brick, but it should be OK just to add a link to `BaseRecurrent` and `@recurrent` since many people use blocks for sequence generation tasks.

File: docs/create_your_own_brick.rst [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+Let's say that you want to create a brick from scracth, simply inheriting
+from :class:`.Brick`, then you should consider overwriting the following
+methods (strictly speaking, all these methods are optional):
+
+* :meth:`.Brick.__init__`: you should pass by argument the attributes of your
+  brick. It is also in this method that you should create the potential
+  "children bricks" that belongs to your brick (in that case, you have to put
+  the children bricks into ``self.children``. The initialiazation of the
+  attributes can be lazy as described later in the tutorial.
```
Comment (dmitriy-serdyuk) [inline docs/create_your_own_brick.rst]: I think, you need to mention that the brick can call application methods of other brick only if it is a child brick.

File: docs/create_your_own_brick.rst [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+* :meth:`you_decide_which_name`: you need to implement a method that actually
+  implements the operation of the brick, taking as arguments the inputs
+  of the brick and returning its outputs. It can have any name and for simple
+  bricks is often named ``apply``. You should decorate it with the
+  :func:`.application` decorator, as explained in the next section.
+* :meth:`.Brick._allocate`: you should implement this method if your brick
+  needs to allocate its parameters.
```
Comment (dmitriy-serdyuk) [inline docs/create_your_own_brick.rst]: We have an agreement, that when bricks are allocated, the parameters contain nans.

**Comment by dmitriy-serdyuk** [comment]
Sorry for pickiness. I just realized that we should explain how to use `get_dim` and lazy constructors in `_push_allocation_config` method. I think we need a brick has two children and the first child's dimension depends on the second. Are you following me?

**Comment by adbrebs** [comment]
@dmitriy-serdyuk I have modified the tutorial according to your comments except the last one (lazy), which I don't really know how to do. Could you please point out an example?

**Comment by dmitriy-serdyuk** [comment]
Usually, `get_dim` is used inside `_push_allocation_config`, see for example `BaseSequenceGenerator._push_allocation_config` class: it assigns dimensions for one child depending on demensions of other. It works nicely with lazy initialization, a user may create a brick and do not specify some dimentions and they will be assigned by a parent brick.

I suggest to create one more brick `DoubleParallel` which applies two parallel-like bricks consequently like

```
class DoubleParallel:
    def __init__(transformations):
         self.transformations = transformations
    def _push_allocation_config():
        # assign output dimension of transformation 0 to the input 
        # dimension of transformation 1
    @application
    def apply(x, y):
        outs = self.transformations[0].apply(x, y)
        return self.transformations[1].apply(outs)

```

**Comment by adbrebs** [comment]
Ok, but what if the two ``transformations`` already have all their attributes initialized? It is not very nice to assign them a second time (actually this might lead to quiet bugs).

Furthermore ``DoubleParallel`` would only work with children bricks that have ``input_dim`` and ``output_dim`` attributes.

**Comment by dmitriy-serdyuk** [comment]
Anyway, the attributes should be the same. I guess one may check if they are initialized and not the same and raise an error. But it is overcomplicated for this tutorial.

You shouldn't use `input/output_dim`. What you need to demostrate is how to extract names of a child brick inputs/outputs using properties `apply.inputs`, `apply.outputs` and how to get dimension information with `get_dim(name)`.

File: docs/create_your_own_brick.rst [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+   >>>
+   >>> f1 = theano.function([x1, x2], [output1, output2])
+   >>> f1(x1_mat, x2_mat) # doctest: +ELLIPSIS
+   [array([[ 60.,  60.]]...), array([[ 40.],
+          [ 40.]]...)]
+
+One can also create the brick using :class:`Linear` children bricks, which
+
+    >>> class ParallelLinear2(Initializable):
+    ...     def __init__(self, input_dim1, input_dim2, output_dim1, output_dim2,
```
Comment (ASalvail) [inline docs/create_your_own_brick.rst]: `@lazy` decorator missing?

Comment (adbrebs) [inline docs/create_your_own_brick.rst]: Thanks for pointing this out!

File: docs/create_your_own_brick.rst [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+    ...     @application(inputs=['input1_', 'input2_'], outputs=['output1',
+    ...         'output2'])
+    ...     def apply(self, input1_, input2_):
+    ...         output1 = self.linear1.apply(input1_)
+    ...         output2 = self.linear2.apply(input2_)
+    ...         return output1, output2
+    ...
+    ...     def get_dim(self, name):
+    ...         if name in ['input1_', 'output1']:
+    ...             return self.linear1.get_dim(name)
```
Comment (adbrebs) [inline docs/create_your_own_brick.rst]: @dmitriy-serdyuk this does not make sense.

**Comment by adbrebs** [comment]
@dmitriy-serdyuk 
Something like that?
```python
class DoubleSequential:
    """
    Two sequential Linear bricks.
    """
    def __init__(self, brick1, brick2):
        self.brick1 = brick1
        self.brick2 = brick2

    def _push_allocation_config(self):
        self.brick2.input_dim = self.brick1.get_dim('input_')

    @application(inputs=['input_'], outputs=['output'])
    def apply(self, x):
        out = self.brick1.apply(x)
        return self.brick2.apply(out)
```
I still don't really understand how to do it and the benefits.

**Comment by rizar** [comment]
@adbrebs, `self.brick1.get_dim(self.brick1.apply.inputs[0])` would be completely interface-independent. But in this particular case it does not help much, because one must know interface of `brick2` to set its configuration. For this example a reasonable requirement would be that both `brick1` and `brick2` support `blocks.bricks.Feedforward` interface. With such a requirement `input_dim` and `output_dim` requests become legitimate.

**Comment by adbrebs** [comment]
Ok thanks @rizar.
@dmitriy-serdyuk I have added the ``DoubleSequential`` brick as you wanted.

File: docs/create_your_own_brick.rst [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+* :class:`.Sequence`: a sequence of bricks.
+* :class:`.Initializable`: a brick that defines a same initialiation scheme
+  (weights and biases) for all its children.
+* :class:`.Feedforward`: declares an interface for bricks with one input and
+  one output.
+* :class:`.Linear`: a linear transformation with optional bias. Inherits from
+  :class:`.Initializable` and :class:`.Feedforward`.
+* :class:`.BaseRecurrent`: the base class for recurrent bricks. Check the
+  :doc:`tutorial about rnns</rnn>` for more information.
+* many mores!
```
Comment (dmitriy-serdyuk) [inline docs/create_your_own_brick.rst]: Typo: more

File: docs/create_your_own_brick.rst [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+Let's say that you want to create a brick from scracth, simply inheriting
+from :class:`.Brick`, then you should consider overwriting the following
+methods (strictly speaking, all these methods are optional):
+
+* :meth:`.Brick.__init__`: you should pass by argument the attributes of your
+  brick. It is also in this method that you should create the potential
+  "children bricks" that belongs to your brick (in that case, you have to put
+  the children bricks into ``self.children``). The initialiazation of the
+  attributes can be lazy as described later in the tutorial.
+* :meth:`you_decide_which_name`: you need to implement a method that actually
```
Comment (dmitriy-serdyuk) [inline docs/create_your_own_brick.rst]: Let's call `apply` here? It's easier and often the brick name already explains what it does (like `Tanh.apply` or `Linear.apply`). What do you think?

File: docs/create_your_own_brick.rst [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+
+you_decide_which_name method
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The :meth:`you_decide_which_name` method listed above is probably the most
+important method of your brick because it is the one that actually takes
+theano tensors as inputs, process them and return output tensors. You should
+decorate it with the :func:`.application` decorator, which names variables
+and register auxiliary variables of the operation you implement.
+It is used as followed (:meth:`you_decide_which_name` is named :meth:`apply`):
```
Comment (dmitriy-serdyuk) [inline docs/create_your_own_brick.rst]: typo: as follows

File: docs/create_your_own_brick.rst [hunk 12]
```diff
@@ -0,0 +1,364 @@
...
+argument of your ``apply`` function, as shown in this example:
+
+    >>> class Foo(Brick): # doctest: +SKIP
+    ...     @application
+    ...     def apply(self, x, application_call):
+    ...         application_call.add_auxiliary_variable(x.mean())
+    ...         return x + 1
+
+You can retrieve these annotated variables as usual with the computational
+graph.
```
Comment (dmitriy-serdyuk) [inline docs/create_your_own_brick.rst]: Can you add links to `ComputationGraph` and `VariableFilter` classes here?

File: docs/create_your_own_brick.rst [hunk 13]
```diff
@@ -0,0 +1,364 @@
...
+    ...     def __init__(self, brick1, brick2, **kwargs):
+    ...         super(Feedforward, self).__init__(**kwargs)
+    ...         self.brick1 = brick1
+    ...         self.brick2 = brick2
+    ...         self.input_dim = brick1.input_dim
+    ...         self.output_dim = brick2.output_dim
+    ...         self.children = [self.brick1, self.brick2]
+    ...
+    ...     def _push_allocation_config(self):
+    ...         self.brick2.input_dim = self.brick1.get_dim('input_')
```
Comment (dmitriy-serdyuk) [inline docs/create_your_own_brick.rst]: It should be `get_dim('output')`.

File: docs/create_your_own_brick.rst [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+    ...     def apply(self, x):
+    ...         return self.brick2.apply(self.brick1.apply(x))
+
+Note how ``get_dim`` is used to retrieve the ``input_dim`` of ``brick1``. You
+can now create a DoubleSeuential brick as follows.
+
+    >>> brick1 = Linear(input_dim=3, output_dim=2)
+    >>> brick2 = Linear(output_dim=4)
+    >>>
+    >>> seq = DoubleSequential(brick1, brick2)
```
Comment (dmitriy-serdyuk) [inline docs/create_your_own_brick.rst]: It worth adding initialization and demonstrate that the input dimension of brick2 was set:
```
seq.initialize()
brick2.input_dim
```

File: docs/create_your_own_brick.rst (hunk 10 above)
Comment (adbrebs) [inline docs/create_your_own_brick.rst]: Ok

File: docs/create_your_own_brick.rst (hunk 4 above)
Comment (adbrebs) [inline docs/create_your_own_brick.rst]: Sorry, I don't get it.

File: docs/create_your_own_brick.rst (hunk 8 above)
Comment (adbrebs) [inline docs/create_your_own_brick.rst]: up

File: docs/create_your_own_brick.rst (hunk 4 above)
Comment (dmitriy-serdyuk) [inline docs/create_your_own_brick.rst]: You had `W1` and `W2` properties earlier. Now it's okay.

File: docs/create_your_own_brick.rst (hunk 8 above)
Comment (dmitriy-serdyuk) [inline docs/create_your_own_brick.rst]: This seems totally fine. If the user asked for dimensions of the first brick, ask brick1, otherwise ask brick2.

File: docs/create_your_own_brick.rst [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+  of the brick and returning its outputs. It can have any name and for simple
+  bricks is often named ``apply``. You should decorate it with the
+  :func:`.application` decorator, as explained in the next section. If you
+  design a recurrent brick, you should instead decorate it with the
+  :func:`.recurrent` decorator as explained in the
+  :doc:`tutorial about rnns</rnn>`.
+* :meth:`.Brick._allocate`: you should implement this method to allocate the
+  shared variables (often representing parameters) of the brick. In blocks,
+  by convention, the built-in bricks allocate their shared variables with nan
+  values and we recommand you to do the same.
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: recommend

File: docs/create_your_own_brick.rst [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+* :meth:`.Brick._initialize`: you should implement this method to initialize
+  the shared variables of your brick. This method is called after the
+  allocation.
+* :meth:`.Brick._push_allocation_config`: you should consider overwriting
+  this method if you want to allocate the children bricks in a specific way.
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: I would say "... if you want to change configuration of the children bricks before they allocate their parameters"

File: docs/create_your_own_brick.rst [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+* :meth:`.Brick._push_initialization_config`: you should consider
+  overwriting this method if you want to initialize the children bricks in a
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: The same rephrasal as for `_push_allocation_config` is proposed. Also, I would elaborate on what `Initializable` does (propagates `weights_init` and `biases_init`).

File: docs/create_your_own_brick.rst [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+  specific way. You should inherit from :class:`.Initializable` to initialize
+  the potential children bricks recursively.
+* :meth:`.Brick.get_dim`: implementing this function is useful if you want
+  to provide a simple way to get the dimensions of the inputs and outputs of
+  the brick.
+
+If you want to inherit from a specific brick, check its docstring to
+identify the particular methods to overwrite and the attributes to define.
+
+apply method
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: Application methods?

File: docs/create_your_own_brick.rst [hunk 19]
```diff
@@ -0,0 +1,371 @@
...
+    ...         ...
+    ...         return something
+
+In the case above, it will automatically label the theano tensor variable
+``input1`` to ``Foo_apply_input1``, idem for ``input2`` and the output of the
+method.
+
+Under the hood, the ``@application`` decorator creates an object of class
+:class:`.Application`, named ``apply``, which becomes an attribute of the
+brick.
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: I would emphasize that it is an attribute of the brick class, not brick object.

File: docs/create_your_own_brick.rst [hunk 20]
```diff
@@ -0,0 +1,371 @@
...
+
+Application properties
+""""""""""""""""""""""
+
+In the previous examples, variables were named with strings. If you need to
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: Variable names are always strings, the only difference is that sometimes they can not be specified at the time when the class is defined, but only when the object is created. This subsection has to be rewritten.

File: docs/create_your_own_brick.rst [hunk 21]
```diff
@@ -0,0 +1,371 @@
...
+    ...         ...
+    ...     @apply.property('inputs')
+    ...     def apply_inputs(self):
+    ...         # Note that you can use any python code to define the name
+    ...         return self.fancy_name
+
+Using application calls
+"""""""""""""""""""""""
+
+You can also annotate specific variables by passing ``application_call`` as
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: The purpose of providing `application_call` is not "to annotate specific variables", but to allow adding auxiliary variables to the computation graph.

File: docs/create_your_own_brick.rst [hunk 22]
```diff
@@ -0,0 +1,371 @@
...
+batch inputs and multiply them respectively by two matrices, resulting in two
+outputs.
+
+The first step is to identify which brick to inherit from. Clearly we are
+implementing a variant of the :class:`.Linear` brick. Contrary to
+:class:`.Linear`, ours has two inputs and two outputs, which means that we can
+not inherit from :class:`.Feedforward`, which requires a single input and a
+single output. Our brick will have to manage two shared variables
+representing the matrices to multiply the inputs with and thus, inheriting from
+:class:`.Initializable` makes perfectly sense as we will let the user decide
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: Inheriting from `Initializable` just adds `weights_init` and `biases_init` in this case, I am not sure I understand what "let the user decide which initialization scheme to use" means.

File: docs/create_your_own_brick.rst [hunk 23]
```diff
@@ -0,0 +1,371 @@
...
+    ...         return output1, output2
+    ...
+    ...     def get_dim(self, name):
+    ...         if name in ['input1_', 'output1']:
+    ...             return self.linear1.get_dim(name)
+    ...         if name in ['input2_', 'output2']:
+    ...             return self.linear2.get_dim(name)
+    ...         super(ParallelLinear2, self).get_dim(name)
+
+You can test this new version as follows:
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: Here we should definitely pay attention to the fact that `weights_init` and `biases_init` are propagated behind the scenes to the linear children, and that this is in effect the main reason to use `Initializable`.

**Comment by rizar** [comment]
@dmitriy-serdyuk asked me to take a look. I like the tutorial a lot, but in some cases it deviates from the truth a little bit.

**Comment by adbrebs** [comment]
Yes I'm not yet very familiar with many blocks features, so I am not always sure I cover every points very well.

**Comment by adbrebs** [comment]
@dmitriy-serdyuk @rizar Thanks for reviewing, it should be better now.

**Comment by adbrebs** [comment]
I don't understand why these tests fail: ``tests.bricks.test_conv.FunctionTestCase (test_convolutional_sequence)``

**Comment by rizar** [comment]
Sorry, out builds currently fail and we are trying to figure out why, see #836 

But since you can be sure that your PR does not break the convolutional bricks, you don't have to worry about them :)

**Comment by adbrebs** [comment]
No risk that the documentation breaks the conv bricks ;).
@dmitriy-serdyuk is this good enough for merging? If yes, I can clean up the commit log before the merge.

File: docs/create_your_own_brick.rst [hunk 24]
```diff
@@ -0,0 +1,387 @@
...
+  :class:`.Initializable` and :class:`.Feedforward`.
+* :class:`.BaseRecurrent`: the base class for recurrent bricks. Check the
+  :doc:`tutorial about rnns</rnn>` for more information.
+* many more!
+
+Let's say that you want to create a brick from scracth, simply inheriting
+from :class:`.Brick`, then you should consider overwriting the following
+methods (strictly speaking, all these methods are optional):
+
+* :meth:`.Brick.__init__`: you should pass by argument the attributes of your
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: Here is a question: I do not remember exactly if we have now a documentation that explains the life-cycle of the brick: creation, allocation of parameters, initialization of parameters, application. One can infer them from this document, but I think it must be stated clearly somewhere...

File: docs/create_your_own_brick.rst [hunk 25]
```diff
@@ -0,0 +1,387 @@
...
+  before they allocate their parameters.
+* :meth:`.Brick._push_initialization_config`: you should consider
+  overwriting this method if you want to change the initialization schemes of
+  the children before they get initialized.
+  If the children bricks need to be initialized with the same scheme, then you
+  should inherit your brick from :class:`.Initializable`, which
+  automatically push the initialization schemes of your brick (provided as
+  arguments ``weights_init`` and ``biases_init`` of the constructor) to the
+  children bricks.
+* :meth:`.Brick.get_dim`: implementing this function is useful if you want
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: I would like to also explain here why using `get_dim` is preferable to accessing the attributes of the brick. The reason is that `get_dim` is the interface that is the same for all bricks, whereas the exact way allocation configuration is stored in a brick object might vary. For instance, `Feedforward` bricks have `input_dim` and `output_dim` attributes, whereas `Parallel` bricks have `input_dims` and `output_dims` (note plural instead of singular). When one wants to change the configuration of a brick, knowing how the attributes are called and what they mean is necessary, but when only read-only access required, `get_dim` should be sufficient. Thus, it was introduced to facilitate writing bricks that use other bricks without making too many assumptions about their interface.

Please let me know if this explanation was helpful and if you can integrate it in a concise way into your tutorial.

File: docs/create_your_own_brick.rst [hunk 26]
```diff
@@ -0,0 +1,387 @@
...
+This tutorial assumes that you are already familiar with
+:doc:`bricks <bricks_overview>` and how to use them from a user point of view.
+
+
+Bricks ingredients and recipe
+-----------------------------
+
+All the bricks in blocks inherit directly or indirectly from the
+:class:`.Brick`. There is already a rich inheritance hierarchy of
+bricks implemented in blocks and thus, you should consider which brick level
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: Blocks, because it is a proper name

File: docs/create_your_own_brick.rst [hunk 27]
```diff
@@ -0,0 +1,387 @@
...
+* :meth:`apply`: you need to implement a method that actually
+  implements the operation of the brick, taking as arguments the inputs
+  of the brick and returning its outputs. It can have any name and for simple
+  bricks is often named ``apply``. You should decorate it with the
+  :func:`.application` decorator, as explained in the next section. If you
+  design a recurrent brick, you should instead decorate it with the
+  :func:`.recurrent` decorator as explained in the
+  :doc:`tutorial about rnns</rnn>`.
+* :meth:`.Brick._allocate`: you should implement this method to allocate the
+  shared variables (often representing parameters) of the brick. In blocks,
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: **B**locks

File: docs/create_your_own_brick.rst [hunk 28]
```diff
@@ -0,0 +1,387 @@
...
+  allocation.
+* :meth:`.Brick._push_allocation_config`: you should consider overwriting
+  this method if you want to change configuration of the children bricks
+  before they allocate their parameters.
+* :meth:`.Brick._push_initialization_config`: you should consider
+  overwriting this method if you want to change the initialization schemes of
+  the children before they get initialized.
+  If the children bricks need to be initialized with the same scheme, then you
+  should inherit your brick from :class:`.Initializable`, which
+  automatically push the initialization schemes of your brick (provided as
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: push*es*

File: docs/create_your_own_brick.rst [hunk 29]
```diff
@@ -0,0 +1,387 @@
...
+and register auxiliary variables of the operation you implement.
+It is used as follows:
+
+    >>> class Foo(Brick): # doctest: +SKIP
+    ...     @application(inputs=['input1', 'input2'], outputs=['output'])
+    ...     def apply(self, input1, input2)
+    ...         ...
+    ...         return something
+
+In the case above, it will automatically label the theano tensor variable
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: Explanation should be accompanied by a code sample which shows what exactly you mean. Also, we should mention here not only how `.name` is changed, but also `.tag.names` and `.tag.roles`.

File: docs/create_your_own_brick.rst [hunk 30]
```diff
@@ -0,0 +1,387 @@
...
+``input1`` to ``Foo_apply_input1``, idem for ``input2`` and the output of the
+method.
+
+Under the hood, the ``@application`` decorator creates an object of class
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: Also, a small code fragment would be helpful (which queries `Foo.apply` and shows what kind of object it is).

File: docs/create_your_own_brick.rst [hunk 31]
```diff
@@ -0,0 +1,387 @@
...
+of your ``apply`` function, as shown in this example:
+
+    >>> class Foo(Brick): # doctest: +SKIP
+    ...     @application
+    ...     def apply(self, x, application_call):
+    ...         application_call.add_auxiliary_variable(x.mean())
+    ...         return x + 1
+
+You can retrieve these annotated variables as usual with the computational
+graph :class:`.ComputationGraph` and filters :class:`.VariableFilter`.
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: Please add an example and please say one sentence why one might need auxiliary variables at all.

File: docs/create_your_own_brick.rst [hunk 32]
```diff
@@ -0,0 +1,387 @@
...
+two :class:`.Feedforward` bricks for which you do not need to
+specify the ``input_dim`` of ``brick2`` directly at its creation.
+
+    >>> class DoubleSequential(Feedforward):
+    ...     """Two sequential Feedforward bricks."""
+    ...     def __init__(self, brick1, brick2, **kwargs):
+    ...         super(Feedforward, self).__init__(**kwargs)
+    ...         self.brick1 = brick1
+    ...         self.brick2 = brick2
+    ...         self.input_dim = brick1.input_dim
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: That's actually an example of a bad behaviour, because `brick1` might not have `input_dim` at that stage. You can see `Sequence` for a more robust example of implementing `Feedforward` interface: https://github.com/mila-udem/blocks/blob/master/blocks/bricks/__init__.py#L641

File: docs/create_your_own_brick.rst [hunk 33]
```diff
@@ -0,0 +1,387 @@
...
+    ... def __init__(self, attr1, attr1)
+    ...     ...
+
+This allows the user to specify ``attr1`` and ``attr2`` after the creation of 
+the brick. For example, the following ``DoubleSequential`` brick is composed of
+two :class:`.Feedforward` bricks for which you do not need to
+specify the ``input_dim`` of ``brick2`` directly at its creation.
+
+    >>> class DoubleSequential(Feedforward):
+    ...     """Two sequential Feedforward bricks."""
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: The word 'sequential' is a bit confusing here. I would rather call the class `ChainOfTwoFeedforward`, this is lengthy but to the point.

File: docs/create_your_own_brick.rst [hunk 34]
```diff
@@ -0,0 +1,387 @@
...
+Note how ``get_dim`` is used to retrieve the ``input_dim`` of ``brick1``. You
+can now use a ``DoubleSeuential`` brick as follows.
+
+    >>> brick1 = Linear(input_dim=3, output_dim=2, use_bias=False,
+    ...                 weights_init=Constant(2))
+    >>> brick2 = Linear(output_dim=4, use_bias=False, weights_init=Constant(2))
+    >>>
+    >>> seq = DoubleSequential(brick1, brick2)
+    >>> seq.initialize()
+    >>> brick2.input_dim
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: I would appreciate if you also show here that setting `weights_init` for `DoubleSequential` would have the same effect as setting it for each of the bricks.

Comment (rizar) [inline docs/create_your_own_brick.rst]: Oh I see, you kind of have it in your next example, and you don't inherit from `Initializable` here. All right!

File: docs/create_your_own_brick.rst [hunk 35]
```diff
@@ -0,0 +1,387 @@
...
+    ...
+    ...     def __allocate(self, input_dim, output_dim, number):
+    ...         W = shared_floatx_nans((input_dim, output_dim),
+    ...                                name='W'+number)
+    ...         add_role(W, WEIGHT)
+    ...         self.parameters.append(W)
+    ...         self.add_auxiliary_variable(W.norm(2), name='W'+number+'_norm')
+    ...
+    ...     def _allocate(self):
+    ...         self.__allocate(self.input_dim1, self.output_dim1, '1')
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: `_allocate`?

File: docs/create_your_own_brick.rst [hunk 36]
```diff
@@ -0,0 +1,387 @@
...
+   ...                         dtype=theano.config.floatX)
+   >>> x2_mat = 4 * numpy.ones((batch_size2, input_dim2),
+   ...                         dtype=theano.config.floatX)
+   >>>
+   >>> x1 = theano.tensor.matrix('x1')
+   >>> x2 = theano.tensor.matrix('x2')
+   >>> parallel1 = ParallelLinear(input_dim1, input_dim2, output_dim1,
+   ...                            output_dim2, weights_init=Constant(2))
+   >>> parallel1.initialize()
+   >>> # The weights_init initialization scheme is pushed to the children
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: This statement should be illustrated by actually checking `weight_init` of the children.

Comment (rizar) [inline docs/create_your_own_brick.rst]: No, hold on, you don't  have children here! Then there is no pushing...

File: docs/create_your_own_brick.rst [hunk 37]
```diff
@@ -0,0 +1,387 @@
...
+    ...         if name in ['input2_', 'output2']:
+    ...             return self.linear2.get_dim(name)
+    ...         super(ParallelLinear2, self).get_dim(name)
+
+You can test this new version as follows:
+
+   >>> parallel2 = ParallelLinear2(input_dim1, input_dim2, output_dim1,
+   ...                             output_dim2, weights_init=Constant(2))
+   >>> parallel2.initialize()
+   >>> output1, output2 = parallel2.apply(x1, x2)
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: Ok, and you here you should actually show that initialization scheme pushing took place.

File: docs/create_your_own_brick.rst (hunk 24 above)
Comment (adbrebs) [inline docs/create_your_own_brick.rst]: Could we point to the docstring of ``Brick``?

File: docs/create_your_own_brick.rst (hunk 32 above)
Comment (adbrebs) [inline docs/create_your_own_brick.rst]: Yes I agree and actually I don't see how lazy initialization can be "robust" in general. In the case of ``Sequence``, don't we have the same problem if the ``input_dim`` of the first child is not defined?

File: docs/create_your_own_brick.rst (hunk 35 above)
Comment (adbrebs) [inline docs/create_your_own_brick.rst]: __allocate is another function with a different signature

**Comment by adbrebs** [comment]
@rizar I have updated the tutorial to take into account most of your comments.

File: docs/create_your_own_brick.rst [hunk 38]
```diff
@@ -0,0 +1,419 @@
...
+    >>>
+    >>> from blocks.bricks import Brick, Initializable, Linear, Feedforward
+    >>> from blocks.bricks.base import lazy, application
+    >>> from blocks.bricks.parallel import Parallel
+    >>> from blocks.initialization import Constant
+    >>> from blocks.roles import add_role, WEIGHT
+    >>> from blocks.utils import shared_floatx_nans
+
+This tutorial explains how to create a custom brick, which is useful if you
+want to factorize a specific several combined operations (which can be made of
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: I would rephrase it as ```if you want to group several specific operations (which can be bricks themselves) into a single...```

File: docs/create_your_own_brick.rst (hunk 24 above)
Comment (rizar) [inline docs/create_your_own_brick.rst]: Yes, this is a good idea.

File: docs/create_your_own_brick.rst (hunk 32 above)
Comment (rizar) [inline docs/create_your_own_brick.rst]: We do have the same problem, but only upon requesting the `input_dim` property, not in the constructor. So you can create a sequence of bricks first and only then set all the dimensions.

File: docs/create_your_own_brick.rst [hunk 39]
```diff
@@ -0,0 +1,419 @@
...
+    ...
+    ...     def _push_allocation_config(self):
+    ...         self.brick2.input_dim = self.brick1.get_dim('output')
+    ...
+    ...     @application
+    ...     def apply(self, x):
+    ...         return self.brick2.apply(self.brick1.apply(x))
+
+Note how ``get_dim`` is used to retrieve the ``input_dim`` of ``brick1``. You
+can now use a ``DoubleSeuential`` brick as follows.
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: It is not `DoubleSequential` anymore, you have changed the naming.

File: docs/create_your_own_brick.rst [hunk 40]
```diff
@@ -0,0 +1,419 @@
...
+    >>> seq.initialize()
+    >>> brick2.input_dim
+    2
+
+
+Example
+-------
+
+For the sake of the tutorial, let's consider a toy operation that takes two
+batch inputs and multiply them respectively by two matrices, resulting in two
```
Comment (rizar) [inline docs/create_your_own_brick.rst]: multiplies

**Comment by adbrebs** [comment]
I have updated the PR. doctest should pass now.

**Comment by rizar** [comment]
Good job, if the tests pass I will merge the PR right after we are finally done with the release (PR #883 ).
